saleor/warehouse/models.py

```diff
@@ -96,10 +113,26 @@ def applicable_for_click_and_collect_no_quantity_check(
             return self._for_channel_click_and_collect(channel_id)
 
         stocks_qs = Stock.objects.filter(
-            product_variant__id__in=lines_qs.values("variant_id"),
-        ).select_related("product_variant")
+            product_variant_id__in=lines_qs.values("variant_id"),
+        )
 
-        return self._for_channel_lines_and_stocks(lines_qs, stocks_qs, channel_id)
+        number_of_variants = (
+            lines_qs.order_by("variant_id").distinct("variant_id").count()
```

Comment:
If we do  `lines_qs.order_by().distinct("variant_id").count()` we should override default ordering without ordering on query.
And futher on, the actual query send to database can be smaller with:
 `lines_qs.order_by().distinct("variant_id").only("variant_id").count()`